libqpdf/QPDF.cc

```diff
@@ -1837,17 +1834,14 @@ QPDF::readObjectAtOffset(
                 offset,
                 "object with ID 0");
         }
-
-        if ((exp_objid >= 0) &&
-            (!((objid == exp_objid) && (generation == exp_generation)))) {
+        if (exp_og.isIndirect() && (exp_og != og)) {
```

Comment:
There are a few changes here that I disagree with but that also expose a weakness of my original implementation. I'm going to push up a conflicting commit that cleans this up, then you can fix your commit. In particular, it is not correct to use isIndirect here -- see later comments plus the comment I have added to the code in my upcoming commit.